apis/src/main/java/org/apache/rocketmq/apis/producer/Producer.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.rocketmq.apis.exception.NetworkTimeoutException;
+import org.apache.rocketmq.apis.exception.TopicDoesNotExistException;
+import org.apache.rocketmq.apis.exception.TransactionCheckerNotSetException;
+import org.apache.rocketmq.apis.message.Message;
+import org.apache.rocketmq.apis.message.MessageView;
+
+/**
+ * Producer is a thread-safe rocketmq client which is used to publish messages.
+ *
+ * <p>On account of network timeout or other reasons, rocketmq producer only promised the at-least-once semantics.
+ * For producer, at-least-once semantics means potentially attempts are made at sending it, messages may be
+ * duplicated but not lost.
+ */
+public interface Producer extends Closeable {
+    /**
+     * Sends a message synchronously.
+     *
+     * <p>This method does not return until it gets the definitive result.
+     *
+     * @param message message to send.
+     * @return the message id assigned to the appointed message.
+     * @throws TopicDoesNotExistException       if the topic of message does not exist.
+     * @throws AuthorisationException           if no permission to send message.
+     * @throws AuthenticationException          if identification could not be recognized by server.
+     * @throws ProducerClosedAlreadyException   if producer is closed already.
+     * @throws MessageTypeDoesNotMatchException if message type does not match with the topic.
+     * @throws NetworkTimeoutException          if encountered network timeout to communicate with server.
+     * @throws NetworkConnectionException       if there is a network connection problem.
+     * @throws PersistenceException             if encountered persistence failure from server.
+     */
+    MessageView send(Message message) throws ClientException;
+
+    /**
+     * Sends a transactional message synchronously.
+     *
+     * @param message     message to send.
+     * @param transaction transaction to bind.
+     * @return the message id assigned to the appointed message.
+     * @throws TopicDoesNotExistException        if the topic of message does not exist.
+     * @throws AuthorisationException            if no permission to send message.
+     * @throws AuthenticationException           if identification could not be recognized by server.
+     * @throws ProducerClosedAlreadyException    if producer is closed already.
+     * @throws MessageTypeDoesNotMatchException  if message type does not match with the topic.
+     * @throws NetworkTimeoutException           if encountered network timeout to communicate with server.
+     * @throws NetworkConnectionException        if there is a network connection problem.
+     * @throws PersistenceException              if encountered persistence failure from server.
+     * @throws TransactionCheckerNotSetException if {@link TransactionChecker} is not set.
+     */
+    MessageView send(Message message, Transaction transaction) throws ClientException;
```

Comment:
Tons of exceptions are thrown, it's daunting.

I suggest we only declare a ClientException with managed code, request-id, message, etc.

These exceptions are unchecked, which means we don't want our users to use these exceptions to control their logic, so a simple ClientException is enough.

Reply:
> These exceptions are unchecked, which means we don't want our users to use these exceptions to control their logic, so a simple ClientException is enough.

First of all, exceptions are inevitable. In fact, about the issue that whether to use checked exception or not, maybe no need to prevent users to do responding operation by the exception thrown deliberately? For example, users just want to count how many times the network timeout exception occurred. .

In addition, about the issue to use error codes or specific exception, actuall they can all solve the problem in theory., here's my concern
 1. It is very easy to abuse error code.
 2. Not all exceptions will have the same parameters, such as requestId, they may only exist in some exceptions.